### ISSUE TITLE
Loss function in SAC

### ISSUE THREAD

**Issue by sweetice** [higgsfield/RL-Adventure-2#11]
In the original SAC paper, the J_pi is:

![image](https://user-images.githubusercontent.com/17249055/53317274-fcc68880-3905-11e9-9cc6-abcca2e03aaf.png)

However, in your implementation, your code is:

```
    policy_loss = (log_prob * (log_prob - log_prob_target).detach()).mean()

    mean_loss = mean_lambda * mean.pow(2).mean()
    std_loss = std_lambda * log_std.pow(2).mean()
    z_loss = z_lambda * z.pow(2).sum(1).mean()

    policy_loss += mean_loss + std_loss + z_loss
```

This is different from the original paper.
I don't understand your implementation. Could you please give me an insights? :)
Thank you!

**Comment by hannesliik** [comment]
The equation is for the derivative of J_pi, ut we needn't calculate the derivative explicitly - PyTorch does it for us.

**Comment by sweetice** [comment]
> The equation is for the derivative of J_pi, but we needn't calculate the derivative explicitly - PyTorch does it for us.

Sorry about that, I have re-edited the problem described.

**Comment by leepengcheng** [comment]
> Sorry about that, I have re-edited the problem described.

-------------
big brother,I also confused by the code. if follow the paper it should be 
```python
policy_loss=log_prob-expected_new_q_value  #when alpha=1
```
I read the implementation of [facebook-horizon-sac](https://github.com/facebookresearch/Horizon/blob/50801113962c8cb145975d67256bb66ab4c229a4/ml/rl/training/sac_trainer.py#L253) , it's same with the paper.

can you explain the meaning of  
    `policy_loss = (log_prob * (log_prob - log_prob_target).detach()).mean()` 
and 
    `policy_loss += mean_loss + std_loss + z_loss` 
for me?
thanks very much

**Comment by G0K0URURI** [comment]
The loss is in the form of KL divergence, which corresponds to equation (10) in the original text.
![image](https://user-images.githubusercontent.com/44106600/133391343-8d63a20a-8ec4-44c0-9695-7f1d9d4969b6.png)
just plug the equation in the definition of KL divergence :-)

**Comment by G0K0URURI** [comment]
This code doesn't apply the reparameterization trick but the likelihood ratio gradient estimator. So this code does not correspond to equation (12).

**Comment by outongyiLv** [comment]
but if I use (12), only change this in policy loss to:
policy_loss=log_prob-expected_new_q_value  #when alpha=1
this will be very pool!!!!

**Comment by G0K0URURI** [comment]
> but if I use (12), only change this in policy loss to: policy_loss=log_prob-expected_new_q_value #when alpha=1 this will be very pool!!!!

**The gradient of eqution (12)  cannot be caculated by backbackpropagating of `policy_loss=log_prob-expected_new_q_value`!!!** 
This is because J_pi contains the expectation of state and action distribution. The expectation is also related to pi, which is ignored in backbackpropagating of `policy_loss=log_prob-expected_new_q_value`. The code uses likelihood ratio gradient estimator to address the issus. For more details about likelihood ratio gradient estimator, please refer to [https://timvieira.github.io/blog/post/2019/04/20/the-likelihood-ratio-gradient/](url)

**Comment by outongyiLv** [comment]
> 

But if I update Policy network first and then update the other two networks, this problem can be solved? well, I know that likelihood ratio gradient estimator is good. But most of the code do not use this method, instead they update Policy first with:policy_loss=log_prob-expected_new_q_value

**Comment by G0K0URURI** [comment]
> > 
> 
> But if I update Policy network first and then update the other two networks, this problem can be solved? well, I know that likelihood ratio gradient estimator is good. But most of the code do not use this method, instead they update Policy first with:policy_loss=log_prob-expected_new_q_value

Only the policy affects the distribution of expectation, so likelihood ratio gradient estimator does not need to be used in Q function and alpha loss function.
